spec/lang/step/operators.md

```diff
@@ -257,19 +257,20 @@ impl<M: Memory> Machine<M> {
 ```rust
 impl<M: Memory> Machine<M> {
     /// Perform a wrapping offset on the given pointer. (Can never fail.)
-    fn ptr_offset_wrapping(&self, ptr: Pointer<M::Provenance>, offset: Int) -> Pointer<M::Provenance> {
+    fn ptr_offset_wrapping(&self, ptr: ThinPointer<M::Provenance>, offset: Int) -> ThinPointer<M::Provenance> {
         ptr.wrapping_offset::<M::T>(offset)
     }
 
     /// Perform in-bounds arithmetic on the given pointer. This must not wrap,
     /// and the offset must stay in bounds of a single allocation.
-    fn ptr_offset_inbounds(&self, ptr: Pointer<M::Provenance>, offset: Int) -> Result<Pointer<M::Provenance>> {
+    // Question: Do we not care about the size here anymore? Or am I missing something?
```

Comment:
The intrinsic requires a sized pointee.

Reply:
Sure, but actually I am more confused about the actual signed_dereferencable function, because what it actually checks is that the range between the two pointers is within the allocation, not that the range from the new pointer to the end of the object is. To me it seems like this should instead be implemented as checking dereferencability from the newly offset pointer.

Reply:
When you offset a pointer by 10, then the old pointer needs to be dereferenceable for 10 bytes.

When you offset it by -10, it needs to be dereferenceable for -10 bytes.

We could equivalently check whether the newly offset pointer is dereferenceable for `-offset` bytes but that only seems more confusing to me.

Reply:
I guess we don't care (yet) about the pointee object. So if the pointer points to a 16 byte struct, and it is offset, so that only 8 bytes fit in the same allocation, then this is not UB yet, only when loading from it?

Reply:
Correct. Raw pointers do not have to point to valid memory, so a pointer to `[u8; 16]` that points to 4 bytes is entirely fine.

However, [`offset`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.offset) has the constraint that the offset operation itself must stay inside the bounds of the allocation, and that's what we are checking here.